src/shims/native_lib/mod.rs

```diff
@@ -267,6 +270,189 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                // as uninit in Miri, but all bytes returned by `MiriAllocBytes` are
+                // initialised.
+                unsafe {
+                    Box::from(std::slice::from_raw_parts(
+                        alloc_ptr.add(ofs),
+                        mplace.layout.size.bytes_usize(),
+                    ))
+                }
+            }
+            either::Either::Right(imm) => {
+                let (first, maybe_second) = imm.to_scalar_and_meta();
+                // If a scalar is a pointer, then expose its provenance.
+                if let interpret::Scalar::Ptr(p, _) = first {
+                    // This relies on the `expose_provenance` in the `visit_reachable_allocs` callback
+                    // below to expose the actual interpreter-level allocation.
+                    this.expose_and_warn(Some(p.provenance), tracing)?;
+                }
+
+                // Turn the scalar(s) into u128s so we can write their bytes
+                // into the buffer.
+                let (sc_int_first, sz_first) = {
+                    let sc = first.to_scalar_int()?;
+                    (sc.to_bits_unchecked(), sc.size().bytes_usize())
+                };
+                let (sc_int_second, sz_second) = match maybe_second {
+                    MemPlaceMeta::Meta(sc) => {
+                        // Might also be a pointer.
+                        if let interpret::Scalar::Ptr(p, _) = first {
+                            this.expose_and_warn(Some(p.provenance), tracing)?;
+                        }
+                        let sc = sc.to_scalar_int()?;
+                        (sc.to_bits_unchecked(), sc.size().bytes_usize())
+                    }
+                    MemPlaceMeta::None => (0, 0),
+                };
+                let sz = imm.layout.size.bytes_usize();
+                // TODO: Is this actually ok? Seems like the only way to figure
+                // out how the scalars are laid out relative to each other.
+                let align_second = match imm.layout.backend_repr {
+                    rustc_abi::BackendRepr::Scalar(_) => 1,
+                    rustc_abi::BackendRepr::ScalarPair(_, sc2) => sc2.align(this).bytes_usize(),
+                    _ => unreachable!(),
+                };
+                // How many bytes to skip between scalars if necessary for alignment.
+                let skip = sz_first.next_multiple_of(align_second).strict_sub(sz_first);
+
+                let mut bytes: Box<[u8]> = (0..sz).map(|_| 0u8).collect();
+
+                // Copy over the bytes in an endianness-agnostic way. Since each
+                // scalar may be up to 128 bits and write_target_uint doesn't
+                // give us an easy way to do multiple writes in a row, we
+                // adapt its logic for two consecutive writes.
+                let mut bytes_wr = bytes.as_mut();
+                match this.data_layout().endian {
+                    rustc_abi::Endian::Little => {
```

Comment:
You should *not* be doing your own endianess handling. `write_target_uint` can do that for you.